internal/lm/fabric.go

```diff
@@ -0,0 +1,156 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package lm
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math/rand" // nolint:gosec
+	"net"
+	"os"
+	"sort"
+	"strings"
+
+	spec "github.com/NVIDIA/k8s-device-plugin/api/config/v1"
+	"github.com/NVIDIA/k8s-device-plugin/internal/resource"
+
+	"github.com/google/uuid"
+	"k8s.io/klog/v2"
+)
+
+func newImexDomainLabeler(config *spec.Config, devices []resource.Device) (Labeler, error) {
+	if config.Flags.GFD.ImexNodesConfigFile == nil {
+		// No imex config file, return empty labels
+		return empty{}, nil
+	}
+
+	imexConfigFile, err := os.Open(*config.Flags.GFD.ImexNodesConfigFile)
+	if os.IsNotExist(err) {
+		// No imex config file, return empty labels
+		return empty{}, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to open imex config file: %v", err)
+	}
+	defer imexConfigFile.Close()
+
+	imexDomainID, err := getImexDomainID(imexConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterUUID, cliqueID, err := getFabricIds(devices)
+	if err != nil {
+		return nil, err
+	}
+	if clusterUUID == "" || cliqueID == "" {
+		return empty{}, nil
+	}
+
+	labels := Labels{
+		"nvidia.com/gpu.clique":      clusterUUID + "-" + cliqueID,
+		"nvidia.com/gpu.imex-domain": strings.Join([]string{imexDomainID, cliqueID}, "-"),
+	}
+
+	return labels, nil
+}
+
+func getFabricIds(devices []resource.Device) (string, string, error) {
+	uniqueClusterUUIDs := make(map[string][]int)
+	uniqueCliqueIDs := make(map[string][]int)
+	for i, device := range devices {
+		isFabricAttached, err := device.IsFabricAttached()
+		if err != nil {
+			return "", "", fmt.Errorf("error checking imex capability: %v", err)
+		}
+		if !isFabricAttached {
+			continue
+		}
+
+		clusterUUID, cliqueID, err := device.GetFabricIds()
+		if err != nil {
+
+			return "", "", fmt.Errorf("error getting fabric IDs: %w", err)
+		}
+
+		uniqueClusterUUIDs[clusterUUID] = append(uniqueClusterUUIDs[clusterUUID], i)
+		uniqueCliqueIDs[cliqueID] = append(uniqueCliqueIDs[cliqueID], i)
+	}
+
+	if len(uniqueClusterUUIDs) > 1 {
+		klog.Warningf("Cluster UUIDs are different: %v", uniqueClusterUUIDs)
+		return "", "", nil
+	}
+
+	if len(uniqueCliqueIDs) > 1 {
+		klog.Warningf("Clique IDs are different: %v", uniqueCliqueIDs)
+		return "", "", nil
+	}
+
+	for clusterUUID := range uniqueClusterUUIDs {
+		for cliqueID := range uniqueCliqueIDs {
+			return clusterUUID, cliqueID, nil
+		}
+	}
+	return "", "", nil
+}
+
+func getImexDomainID(r io.Reader) (string, error) {
```

Comment:
Let's add a description of what this does. Should we pass the cliqueID here?

Reply:
Imex-domains are orthogonal to cliques. You might have multiple cliques in a single IMEX domain or the same clique across multiple IMEX domains.

Reply:
`nvidia.com/gpu.imex-domain` is one label, `nvidia.com/gpu.clique` is another. The label we are dropping is `nvidia.com/gpu.clusteruuid` (which is now being prepended to the value of the `nvidia.com/gpu.clique`).